semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/cfgdominator/CfgDominator.scala

```diff
@@ -0,0 +1,114 @@
+package io.shiftleft.semanticcpg.passes.cfgdominator
+
+class CfgDominator[Node](adapter: CfgAdapter[Node]) {
+
+  /**
+    * Calculates the immediate dominators of all CFG nodes reachable from cfgEntry.
+    * Since the cfgEntry does not have an immediate dominator, it has no entry in the
+    * return map.
+    *
+    * The used algorithm is from: "A Simple, Fast Dominance Algorithm" from
+    * "Keith D. Cooper, Timothy J. Harvey, and Ken Kennedy".
+    */
+  def calculate(cfgEntry: Node): Map[Node, Node] = {
+    val UNDEFINED = -1
+    val postOrderNumbering = createPostOrderNumbering(cfgEntry)
+    val indexOf = postOrderNumbering.withDefaultValue(UNDEFINED) // Index of each node into dominators array.
+    // We use withDefault because unreachable/dead
+    // code nodes are not numbered but may be touched
+    // as predecessors of reachable nodes.
+    val nodesInReversePostOrder = postOrderNumbering.toList // Does not contain entry.
+      .sortBy { case (node, index) => -index }
+      .map { case (node, index) => node }
+      .filter { _ != cfgEntry }
+    val dominators = Array.fill(indexOf.size)(UNDEFINED)
+
+    // Used in places where index might be UNDEFINED.
+    def saveDominators(index: Int): Int = {
+      if (index != UNDEFINED) {
+        dominators(index)
+      } else {
+        UNDEFINED
+      }
+    }
+
+    dominators(indexOf(cfgEntry)) = indexOf(cfgEntry)
+
+    var changed = true
+    while (changed) {
+      changed = false
+      nodesInReversePostOrder.foreach { node =>
+        val firstNotUndefinedPred = adapter
+          .predecessors(node)
+          .find { predecessor =>
+            saveDominators(indexOf(predecessor)) != UNDEFINED
+          }
+          .get
+
+        var newImmediateDominator = indexOf(firstNotUndefinedPred)
+        adapter.predecessors(node).foreach { predecessor =>
+          val predecessorIndex = indexOf(predecessor)
+          if (saveDominators(predecessorIndex) != UNDEFINED) {
+            newImmediateDominator = intersect(dominators, predecessorIndex, newImmediateDominator)
+          }
+        }
+
+        val nodeIndex = indexOf(node)
+        if (dominators(nodeIndex) != newImmediateDominator) {
+          dominators(nodeIndex) = newImmediateDominator
+          changed = true
+        }
+      }
+
+    }
```

Comment:
For the future, we can eliminate the `var` here by replacing:
- L38 with `while (true)` 
- L59 with `continue`
- L62 with `break`